xmtp_mls/src/storage/encrypted_store/user_preferences.rs

```diff
@@ -1,67 +1,93 @@
+use crate::{
+    groups::device_sync::preference_sync::UserPreferenceUpdate, storage::StorageError,
+    subscriptions::LocalEvents, Store,
+};
 
 use super::{
     schema::user_preferences::{self, dsl},
     DbConnection,
 };
 use diesel::prelude::*;
 use rand::{rngs::OsRng, RngCore};
+use tokio::sync::broadcast::Sender;
 
-#[derive(Identifiable, Insertable, Queryable, Debug, Clone, PartialEq, Eq)]
+#[derive(Identifiable, Queryable, AsChangeset, Debug, Clone, PartialEq, Eq, Default)]
 #[diesel(table_name = user_preferences)]
 #[diesel(primary_key(id))]
 pub struct StoredUserPreferences {
     /// Primary key - latest key is the "current" preference
-    pub id: Option<i32>,
+    pub id: i32,
     /// Randomly generated hmac key root
-    pub hmac_key: Vec<u8>,
+    pub hmac_key: Option<Vec<u8>>,
 }
 
-impl Default for StoredUserPreferences {
-    fn default() -> Self {
-        let mut hmac_key = vec![0; 32];
-        OsRng.fill_bytes(&mut hmac_key);
+#[derive(Insertable)]
+#[diesel(table_name = user_preferences)]
+pub struct NewStoredUserPreferences<'a> {
+    hmac_key: Option<&'a Vec<u8>>,
+}
 
-        Self { id: None, hmac_key }
+impl<'a> From<&'a StoredUserPreferences> for NewStoredUserPreferences<'a> {
+    fn from(value: &'a StoredUserPreferences) -> Self {
+        Self {
+            hmac_key: value.hmac_key.as_ref(),
+        }
     }
 }
 
-impl_store!(StoredUserPreferences, user_preferences);
+impl Store<DbConnection> for StoredUserPreferences {
+    fn store(&self, conn: &DbConnection) -> Result<(), StorageError> {
+        conn.raw_query(|conn| {
+            diesel::update(dsl::user_preferences)
+                .set(self)
+                .execute(conn)
+        })?;
+
+        Ok(())
+    }
+}
 
 impl StoredUserPreferences {
     pub fn load(conn: &DbConnection) -> Result<Self, StorageError> {
         let query = dsl::user_preferences.order(dsl::id.desc()).limit(1);
         let mut result = conn.raw_query(|conn| query.load::<StoredUserPreferences>(conn))?;
 
-        let result = match result.pop() {
-            Some(result) => result,
-            None => {
-                // Create a default and store it.
-                let result = Self::default();
-                // TODO: emit an hmac key update event here.
-                result.store(conn)?;
-                result
-            }
-        };
-
-        Ok(result)
+        Ok(result.pop().unwrap_or_default())
     }
 
-    pub fn set_hmac_key(conn: &DbConnection, hmac_key: Vec<u8>) -> Result<(), StorageError> {
+    pub fn new_hmac_key<C>(
+        conn: &DbConnection,
+        local_events: &Sender<LocalEvents<C>>,
+    ) -> Result<Vec<u8>, StorageError> {
         let mut preferences = Self::load(conn)?;
-        // Have the id increment
-        preferences.id = None;
-        preferences.hmac_key = hmac_key;
 
-        preferences.store(conn)?;
-
-        Ok(())
+        let mut hmac_key = vec![0; 32];
+        OsRng.fill_bytes(&mut hmac_key);
```

Comment:
now can use `let hmac_key = xmtp_common::rand_vec::<32>()`

You'll just have to move it from `test.rs` to `lib.rs`, shouldn't need to refactor since test is re-exported at the root